datafusion/physical-plan/src/joins/sort_merge_join.rs

```diff
@@ -991,6 +992,9 @@ impl SMJStream {
             Ordering::Equal => {
                 if matches!(self.join_type, JoinType::LeftSemi) {
                     join_streamed = !self.streamed_joined;
+                    // if the join filter specified there can be references to buffered columns
+                    // so its needed to join them
+                    join_buffered = self.filter.is_some();
```

Comment:
LeftSemi doesn't join buffered side, why we want to do this?

Reply:
Oh, I see. As the filter uses buffered columns, we need to access to it.

Reply:
But the comment doesn't look correct as we don't actually join buffered columns.

Reply:
Hmm, but if there are more rows at buffered side are matched on keys, won't it add additional joined pairs with nulls and buffered rows?

Reply:
I hope the concern right and wrapped it into SQL query
```
query II
select * from (
with
t1 as (
    select 11 a, 12 b union all
    select 11 a, 13 b),
t2 as (
    select 11 a, 12 b union all
    select 11 a, 12 b union all
    select 11 a, 14 b
    )
select t1.* from t1 where exists (select 1 from t2 where t2.a = t1.a and t2.b != t1.b)
) order by 1, 2;
----
11 12
11 13
```
it passes, I can add it to slt file as well

Reply:
I think above test doesn't hit the case I mean.

If last join loop already joined streamed row and buffered rows and the output size reaches batch size, the loop will exit and go to output joined paired.

After that, next join loop gets started again. 

Now the `current_ordering` is `Equal` and `self.streamed_joined` is true, the loop will join nulls and buffered rows. As the left side are nulls, the join filter doesn't work.

This is not caught by the test you posted, because all joined pairs will be output in one batch, I think.

Reply:
Thanks @viirya I'll try to add tests separating the data across multiple batches

Reply:
I'm not able to run such test just because of other SMJ issue not related to this PR:
If the join filter is set and for the same streaming index there are matching rows more or equal to a batch size then the query just stuck. Likely the problem is in polling state and it can be easily reproduced on main branch.

```
  #[tokio::test]
    async fn test_11() -> Result<()> {
        let ctx: SessionContext = SessionContext::new();

        let sql = "set datafusion.optimizer.prefer_hash_join = false;";
        let _ = ctx.sql(sql).await?.collect().await?;

        let sql = "set datafusion.execution.batch_size = 1";
        let _ = ctx.sql(sql).await?.collect().await?;

        let sql = "
        select * from (
        with
        t1 as (
            select 12 a, 12 b
            ),
        t2 as (
            select 12 a, 12 b
            )
            select t1.* from t1 join t2 on t1.a = t2.b where t1.a > t2.b
        ) order by 1, 2;
        ";

        let actual = ctx.sql(sql).await?.collect().await?;


        Ok(())
    }
```

I'll file a separate issue for this one, but perhaps we can go with this PR because potential problem you talking about cannot ever be hit because of the  issue above

Reply:
Thanks for creating the ticket.

I think that current fix to the LeftSemi isn't correct due to the additional joined pairs of nulls and buffered rows. I don't think we should move forward with it because of that.

Reply:
I'm putting the PR to draft until we can check if PR requires modifications to avoid addition join pairs of nulls and buffered rows. It depends on #10491

Reply:
I have checked the case, it doesn't fail, however it produces more rows than expected, looking into this

Reply:
Yea, what I meant is actually it will produce some rows that are not correct results.